476 when i forget the indefinite article

tfnribeiro opened the pull request:
This change introduces a check that takes priority every other hint to check for the case where a user inputs a single word when the expected answer contains 2 words.

Example of the change (expected answer is `best life`):

If the word in the input is an expected word of the output a new hint will be provided:

![image](https://github.com/user-attachments/assets/0fd5a397-7d97-4903-a817-d555767779dc)

However, if the user types anything else then this hint is no longer provided, as the text would not make sense in this case:

![image](https://github.com/user-attachments/assets/4049e9b6-0c26-43de-bc86-085551a8baab)

In cases like the one above, the logic is followed as previously, checking the Levenshtein distance and providing feedback based on that rather than the words themselves. 

If the word is incorrect, it will still report an error because the distance is too far:

![image](https://github.com/user-attachments/assets/21ee64c5-bcf1-40c9-8c33-c3e30fc70cec)

![image](https://github.com/user-attachments/assets/b725e132-ce31-4769-b48e-59ce3f96e557)


The most helpful scenario would probably be in cases where there is a missing article when a language uses form declinations as the following example:

![image](https://github.com/user-attachments/assets/3373d002-7873-4218-9b88-4f02358fd644)

![image](https://github.com/user-attachments/assets/d4a1a709-9fb2-40fe-b6a9-27ffa7f6cf37)

mircealungu commented:
Cool! I've tried to extract the function to a separate file, struggled a bit with a more precise name for it, and then moved it to yet another file in the ./utils/text where it belongs better I think. Please check to see if you agree.

tfnribeiro commented on src/exercises/exerciseTypes/BottomInput.js:
```diff
@@ -125,6 +132,8 @@ export default function BottomInput({
       // we give them a Hint, mainly for audio exercises.
       updatedMessageToAPI = messageToAPI + "H";
       setDistanceToCorrect();
+    } else if (isOneWordCorrect) {
```
I am pretty sure that this won't run as the setter it's only run at the end of the code - line 118 needs to be stored in a temporary value that is used to compute the logic and not the React state.

tfnribeiro commented on src/utils/text/expressions.js:
```diff
@@ -0,0 +1,29 @@
+  return isWordIncluded;
+}
+
+export { isWordIncluded };
```
`export { isWordIncluded, isTextInSentence, isExpression};` is the way we usually export multiple functions?

mircealungu replied:
yes, that was the refactoring engine of WebStorm that did that. I looked at it, it looked strange, and then thought... whatever :) but let's do it our way!

mircealungu commented:
uff. is there still a bug in there? or did I introduce one? I only did refactorings though...

<img width="758" alt="image" src="https://github.com/user-attachments/assets/ee52da5b-04ae-4d3e-9bde-6e9b46581a5c">

in this above exercise I had to type the answer 
- i typed "apply"
- i got as feedback "you are still missing a word"
- then i showed solution

how could this be?

tfnribeiro reviewed (commented):
I think it looks good - though there we cannot use values that are set to do logic in the same place they are set - as the setter only takes effect after the method call. So we need to have a temporary variable , in checkResult to perform the check.

mircealungu commented:
ok, I think it was me to introduce then bug. it's fixed here: https://github.com/zeeguu/web/pull/484/commits/e7a31ba7e94a0e6bec354edf501c2155dec3acd7

tfnribeiro commented:
Yes, I just noticed the word in the for loop matching the word in the parameter bug.  This is how I re-wrote it to be more clear:

```
function isWordIncluded(word, expression) {
  let isWordIncluded = false;
  let wordsInAnswer = expression.split(" ");
  let isInputSingleWord = word.split(" ").length === 1;
  if (isInputSingleWord)
    wordsInAnswer.forEach((wordInAwnser) => {
      if (wordInAwnser === word) isWordIncluded = true;
    });
  return isWordIncluded;
}
```

But, I think what we have now should work too.

mircealungu commented:
> Yes, I just noticed the word in the for loop matching the word in the parameter bug. This is how I re-wrote it to be more clear:
> 
> ```
> function isWordIncluded(word, expression) {
>   let isWordIncluded = false;
>   let wordsInAnswer = expression.split(" ");
>   let isWordSingleWord= word.split(" ").length === 1;
>   if (isWordSingleWord)
>     wordsInAnswer.forEach((wordInAwnser) => {
>       if (wordInAwnser === word) isWordIncluded = true;
>     });
>   return isWordIncluded;
> }
> ```
> 
> But, I think what we have now should work too.

you're fast :) 
feel free to do another commit if you want.
regarding the `forEach` I learned to use the `each` variable naming pattern in my Smalltalk days... and I find that easier :) but if you want to do another commit with your clarifications, go for it! especially the isSingleWord.

tfnribeiro commented:
I have pushed the change I made, which can be done by using the isExpression util function which makes it ever better I feel. 

In regards to each, I think it's because in python you tend to describe what you are iterating `for word in text` for example, so I have stuck to that pattern of not using a generic each when iterating iterables .
